src/test-run/commands/assertion-executor-factory.js

```diff
@@ -0,0 +1,133 @@
+import { assert, expect } from 'chai';
+import { EventEmitter } from 'events';
+import delay from '../../utils/delay';
+import { ExternalAssertionLibraryError } from '../../errors/test-run';
+import ClientFunctionResultPromise from '../../client-functions/result-promise';
+
+const ASSERTION_DELAY = 200;
+
+export default class AssertionExecutorFactory extends EventEmitter {
+    constructor () {
+        super();
+    }
+
+    _wrapSelectorResultAssertionExecutor (command, executor) {
+        return async () => {
+            var startTime     = new Date();
+            var resultPromise = command.actual;
+            var passed        = false;
+            var reexecuted    = false;
+            var messageSent   = false;
+
+            while (!passed) {
+                command.actual = await resultPromise._reExecute();
+
+                try {
+                    if (reexecuted && !messageSent) {
+                        this.emit('start-assertion-execution', command.options.timeout - (new Date() - startTime));
+                        messageSent = true;
+                    }
+
+                    executor();
+                    passed = true;
+
+                    if (messageSent)
+                        this.emit('end-assertion-execution', true);
+                }
+
+                catch (err) {
+                    if (new Date() - startTime >= command.options.timeout) {
+                        if (messageSent)
+                            this.emit('end-assertion-execution', false);
+
+                        throw err;
+                    }
+
+                    if (!reexecuted)
+                        reexecuted = true;
+
+                    await delay(ASSERTION_DELAY);
```

Comment:
It's become way to complicated. Can we refactor it a bit? Also it would be nice to use vars for time-related stuff with some meaningful name.